src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSTempPECompiler.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.CPS
+{
+    [Export(typeof(ITempPECompiler))]
+    internal class CPSTempPECompiler : ITempPECompiler
+    {
+        public async Task<bool> CompileAsync(IWorkspaceProjectContext context, string outputFileName, string[] filesToInclude, CancellationToken cancellationToken)
+        {
+			if (filesToInclude == null || filesToInclude.Length == 0)
+			{
+				throw new ArgumentException(nameof(filesToInclude), "Must specify some files to compile.");
+			}
+			if (string.IsNullOrWhiteSpace(outputFileName))
+			{
+				throw new ArgumentException(nameof(outputFileName), "Must specify a filename to output to.");
+			}
+
+            var snapshot = ((CPSProject)context).GetProjectSnapshot();
+            // Allow for faster checking because projects could be very large
+            var files = new HashSet<string>(filesToInclude);
+
+            // Remove all files except the ones we care about
+            var documents = snapshot.Documents;
+            foreach (var document in documents)
+            {
+                if (!files.Contains(document.FilePath))
```

Comment:
Consider checking cancellation inside of the loop since there could be hundreds of documents.